File: Contour-Detection-using-OpenCV/python/contour_extraction/contour_extraction.py
"""Compare contour retrieval modes using OpenCV 4 or OpenCV 5."""

from __future__ import annotations

import argparse
from pathlib import Path

import cv2


DEFAULT_INPUT = Path(__file__).resolve().parents[2] / "input" / "custom_colors.jpg"
MODES = {
    "list": cv2.RETR_LIST,
    "external": cv2.RETR_EXTERNAL,
    "ccomp": cv2.RETR_CCOMP,
    "tree": cv2.RETR_TREE,
}


def show_image(title: str, image, display: bool) -> None:
    """Display an image when running interactively."""
    if display:
        cv2.imshow(title, image)
        cv2.waitKey(0)
        cv2.destroyAllWindows()


def run(input_path: Path, output_dir: Path, display: bool = True) -> dict[str, int]:
    """Run the retrieval-mode experiment and return contour counts."""
    image = cv2.imread(str(input_path))
    if image is None:
        raise FileNotFoundError(f"Could not read input image: {input_path}")

    output_dir.mkdir(parents=True, exist_ok=True)
    gray = cv2.cvtColor(image, cv2.COLOR_BGR2GRAY)
    _, threshold_image = cv2.threshold(gray, 150, 255, cv2.THRESH_BINARY)

    counts: dict[str, int] = {}
    for name, mode in MODES.items():
        contours, hierarchy = cv2.findContours(
            threshold_image.copy(), mode, cv2.CHAIN_APPROX_NONE
        )
        rendered = image.copy()
        cv2.drawContours(rendered, contours, -1, (0, 255, 0), 2, cv2.LINE_AA)

        show_image(name.upper(), rendered, display)
        output_path = output_dir / f"contours_retr_{name}.jpg"
        if not cv2.imwrite(str(output_path), rendered):
            raise OSError(f"Could not write output image: {output_path}")

        counts[name] = len(contours)
        hierarchy_shape = None if hierarchy is None else hierarchy.shape
        print(f"{name.upper()}: contours={len(contours)}, hierarchy={hierarchy_shape}")

    return counts


def validate(counts: dict[str, int]) -> None:
    """Check relationships that should hold for the bundled tutorial image."""
    if not all(count > 0 for count in counts.values()):
        raise RuntimeError("Every retrieval mode should find at least one contour")
    if counts["external"] > counts["list"]:
        raise RuntimeError("RETR_EXTERNAL cannot return more contours than RETR_LIST")
    if counts["ccomp"] != counts["tree"]:
        raise RuntimeError("The bundled image should have equal CCOMP and TREE counts")


def parse_args() -> argparse.Namespace:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--input", type=Path, default=DEFAULT_INPUT)
    parser.add_argument("--output-dir", type=Path, default=Path.cwd())
    parser.add_argument("--no-display", action="store_true", help="Skip GUI windows")
    parser.add_argument(
        "--validate", action="store_true", help="Validate bundled-image invariants"
    )
    return parser.parse_args()


if __name__ == "__main__":
    args = parse_args()
    result = run(args.input, args.output_dir, display=not args.no_display)
    if args.validate:
        validate(result)
        print("Validation passed")

File: Contour-Detection-using-OpenCV/cpp/contour_extraction/contour_extraction.cpp
#include <opencv2/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgcodecs.hpp>
#include <opencv2/imgproc.hpp>

#include <array>
#include <filesystem>
#include <iostream>
#include <stdexcept>
#include <string>
#include <vector>

#ifndef TUTORIAL_INPUT_DIR
#define TUTORIAL_INPUT_DIR "../../input"
#endif

namespace fs = std::filesystem;

struct Options {
    fs::path input = fs::path(TUTORIAL_INPUT_DIR) / "custom_colors.jpg";
    fs::path output_dir = fs::current_path();
    bool display = true;
    bool validate = false;
};

Options parse_options(int argc, char** argv) {
    Options options;
    for (int i = 1; i < argc; ++i) {
        const std::string argument = argv[i];
        if ((argument == "--input" || argument == "--output-dir") && i + 1 >= argc) {
            throw std::invalid_argument(argument + " requires a path");
        }
        if (argument == "--input") {
            options.input = argv[++i];
        } else if (argument == "--output-dir") {
            options.output_dir = argv[++i];
        } else if (argument == "--no-display") {
            options.display = false;
        } else if (argument == "--validate") {
            options.validate = true;
        } else {
            throw std::invalid_argument("Unknown argument: " + argument);
        }
    }
    return options;
}

void show_image(const std::string& title, const cv::Mat& image, bool display) {
    if (display) {
        cv::imshow(title, image);
        cv::waitKey(0);
        cv::destroyAllWindows();
    }
}

int main(int argc, char** argv) {
    try {
        const Options options = parse_options(argc, argv);
        const cv::Mat image = cv::imread(options.input.string());
        if (image.empty()) {
            throw std::runtime_error("Could not read input image: " + options.input.string());
        }
        fs::create_directories(options.output_dir);

        cv::Mat gray;
        cv::cvtColor(image, gray, cv::COLOR_BGR2GRAY);
        cv::Mat threshold_image;
        cv::threshold(gray, threshold_image, 150, 255, cv::THRESH_BINARY);

        struct RetrievalMode {
            const char* name;
            int value;
        };
        const std::array<RetrievalMode, 4> modes{{
            {"list", cv::RETR_LIST},
            {"external", cv::RETR_EXTERNAL},
            {"ccomp", cv::RETR_CCOMP},
            {"tree", cv::RETR_TREE},
        }};
        std::array<std::size_t, 4> counts{};

        for (std::size_t index = 0; index < modes.size(); ++index) {
            std::vector<std::vector<cv::Point>> contours;
            std::vector<cv::Vec4i> hierarchy;
            cv::findContours(
                threshold_image.clone(), contours, hierarchy, modes[index].value,
                cv::CHAIN_APPROX_NONE
            );

            cv::Mat rendered = image.clone();
            cv::drawContours(
                rendered, contours, -1, cv::Scalar(0, 255, 0), 2, cv::LINE_AA
            );
            show_image(modes[index].name, rendered, options.display);

            const fs::path output =
                options.output_dir / (std::string("contours_retr_") + modes[index].name + ".jpg");
            if (!cv::imwrite(output.string(), rendered)) {
                throw std::runtime_error("Could not write output image: " + output.string());
            }
            counts[index] = contours.size();
            std::cout << modes[index].name << ": contours=" << contours.size()
                      << ", hierarchy=" << hierarchy.size() << '\n';
        }

        if (options.validate) {
            if (counts[0] == 0 || counts[1] == 0 || counts[2] == 0 || counts[3] == 0 ||
                counts[1] > counts[0] || counts[2] != counts[3]) {
                throw std::runtime_error("Contour retrieval-mode validation failed");
            }
            std::cout << "Validation passed\n";
        }
        return 0;
    } catch (const std::exception& error) {
        std::cerr << "Error: " << error.what() << '\n';
        return 1;
    }
}
